### DISCUSSION TITLE
3D OpenGL Integration Issues

### DISCUSSION THREAD

**Discussion by markusalbrecht-procam** [AvaloniaUI/Avalonia#14213] [General]
Hi there,
we finally managed to get a basis Prototype with OpenGL Integration using the OpenGlControlBase within a 3DControl. The problem is we only get it to Render once it is loaded - when adding new Files the respective events OnOpenGlInit and OnOpenGlRender are not firing. This means we can only load data on initialization directly via the constructor but cannot add further data or reinitialize the 3DContext as there are no mehtods availabel for that.

First Result loading STL data
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/ec68043f-1de8-463f-baf4-06573af900d9)

View3DControl
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/3f403aef-c66e-4245-93a6-6a8d370d8e10)

MainView
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/00ab4fb4-7257-498f-8a06-66475ba046b7)

MainViewModel
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/965d207c-8ed1-4466-915f-45bec54618b6)

Every approach to reload the 3DContext leads to empty results. What would be the best approach to update/rerender the control when exchanging or adding new Models ?

We also might need multiple controls means multiple instances of the 3DContext how to achieve that then ?

**Comment by markusalbrecht-procam** [comment]
Is Silk.Net able to cope with that as it seems we got the re-render functionality there ???
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/cc1f8cb9-5139-47b5-9f7c-26cc5dd2d26a)

**Reply by kekekeks** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/a6918116bdaf3899e28a209e19a7555306d5659c/src/Avalonia.OpenGL/Controls/OpenGlControlBase.cs#L225

**Reply by markusalbrecht-procam** [reply]
So this should update the complete visual tree I RequestNextFrameRendering() guess, I will try it.

**Reply by markusalbrecht-procam** [reply]
@kekekeks Well did not solve the issue, for some reason if we load a new file it is not updated, this is what we do.
The intial load works fine also the manipulation of properties like Yaw/Roll,....
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/f787cd20-8a49-4caa-a7a8-35bbeb6184d8)

On first load OnOpenGlInit and OnOpenGlRender get called and render the loaded Triangles of the STL, I separated the actual functionality into methods in order to call them on reload of a new model:

```csharp
private unsafe void GlInit(GlInterface GL)
{
    // set internal context
    if (GL != null)
        GLInternal = GL;

    if (_points is null || _indices is null)
    {
        return;
    }

    //Info = $"Renderer: {GL.GetString(GL_RENDERER)} Version: {GL.GetString(GL_VERSION)}";

    // Load the source of the vertex shader and compile it.
    _vertexShader = GL.CreateShader(GL_VERTEX_SHADER);
    Console.WriteLine(GL.CompileShaderAndGetError(_vertexShader, VertexShaderSource));

    // Load the source of the fragment shader and compile it.
    _fragmentShader = GL.CreateShader(GL_FRAGMENT_SHADER);
    Console.WriteLine(GL.CompileShaderAndGetError(_fragmentShader, FragmentShaderSource));

    // Create the shader program, attach the vertex and fragment shaders and link the program.
    _shaderProgram = GL.CreateProgram();
    GL.AttachShader(_shaderProgram, _vertexShader);
    GL.AttachShader(_shaderProgram, _fragmentShader);
    const int positionLocation = 0;
    const int normalLocation = 1;
    GL.BindAttribLocationString(_shaderProgram, positionLocation, "aPos");
    GL.BindAttribLocationString(_shaderProgram, normalLocation, "aNormal");
    Console.WriteLine(GL.LinkProgramAndGetError(_shaderProgram));
    CheckError(GL);

    // Create the vertex buffer object (VBO) for the vertex data.
    _vertexBufferObject = GL.GenBuffer();
    // Bind the VBO and copy the vertex data into it.
    GL.BindBuffer(GL_ARRAY_BUFFER, _vertexBufferObject);
    CheckError(GL);
    var vertexSize = Marshal.SizeOf<ModelVertex>();
    fixed (void* pdata = _points)
        GL.BufferData(GL_ARRAY_BUFFER, new nint(_points.Length * vertexSize), new nint(pdata), GL_STATIC_DRAW);

    _indexBufferObject = GL.GenBuffer();
    GL.BindBuffer(GL_ELEMENT_ARRAY_BUFFER, _indexBufferObject);
    CheckError(GL);
    fixed (void* pdata = _indices)
        GL.BufferData(GL_ELEMENT_ARRAY_BUFFER, new nint(_indices.Length * sizeof(uint)), new nint(pdata), GL_STATIC_DRAW);
    CheckError(GL);
    _vertexArrayObject = GL.GenVertexArray();
    GL.BindVertexArray(_vertexArrayObject);
    CheckError(GL);

    GL.VertexAttribPointer(positionLocation, 3, GL_FLOAT, 0, vertexSize, nint.Zero);
    GL.VertexAttribPointer(normalLocation, 3, GL_FLOAT, 0, vertexSize, new nint(12));

    GL.EnableVertexAttribArray(positionLocation);
    GL.EnableVertexAttribArray(normalLocation);
    oglInitiated = true;
}
```
```csharp
 public unsafe void GLRender(GlInterface gl, int fb)
 {
     gl.ClearColor(0.5f, 0.5f, 0.5f, 1.0f);
     gl.Clear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
     gl.Enable(GL_DEPTH_TEST);
     //gl.Disable(GL_CULL_FACE);
     gl.Viewport(0, 0, (int)Bounds.Width, (int)Bounds.Height);

     var GL = gl;

     GL.BindBuffer(GL_ARRAY_BUFFER, _vertexBufferObject);
     GL.BindBuffer(GL_ELEMENT_ARRAY_BUFFER, _indexBufferObject);
     GL.BindVertexArray(_vertexArrayObject);
     GL.UseProgram(_shaderProgram);
     CheckError(GL);
     Matrix4x4 projection = Matrix4x4.CreatePerspectiveFieldOfView((float)(Math.PI / 4), (float)(Bounds.Width / Bounds.Height), 0.01f, 1000.0f);

     Matrix4x4 view = Matrix4x4.CreateLookAt(new Vector3(0.0f, 0.0f, 10.0f + _zoom * 100.0f), new Vector3(), new Vector3(0.0f, 1.0f, 0.0f));
     Matrix4x4 viewProjection = view * projection;

     Matrix4x4 modelRot = Matrix4x4.CreateFromYawPitchRoll(_yaw, _pitch, _roll);
     Matrix4x4 modelScale = Matrix4x4.CreateScale(Math.Max(0.05f, _scale));
     Matrix4x4 modelTranslate = Matrix4x4.CreateTranslation(_offsetX - 25.0f, _offsetY - 25.0f, _offsetZ - 25.0f);

     Matrix4x4 model = modelRot * modelScale * modelTranslate;

     var modelLoc = GL.GetUniformLocationString(_shaderProgram, "uModel");
     var viewProjLoc = GL.GetUniformLocationString(_shaderProgram, "uViewProj");

     var maxYLoc = GL.GetUniformLocationString(_shaderProgram, "uMaxY");
     var minYLoc = GL.GetUniformLocationString(_shaderProgram, "uMinY");

     GL.UniformMatrix4fv(modelLoc, 1, false, &model);
     GL.UniformMatrix4fv(viewProjLoc, 1, false, &viewProjection);
     GL.Uniform1f(maxYLoc, _maxY);
     GL.Uniform1f(minYLoc, _minY);
     CheckError(GL);
     // draw elements if indices are visible
     if (_indices != null)
         GL.DrawElements(GL_TRIANGLES, _indices.Length, GL_UNSIGNED_INT, nint.Zero);

     CheckError(GL);

     float curElapsedSeconds = (float)St.Elapsed.TotalSeconds;
     float dT = curElapsedSeconds - lastElapsedTime;
     if (dT > 0.166)
     {
         lastElapsedTime = curElapsedSeconds;
         RequestNextFrameRendering();
     }
 }
```

So maybe the integration of using the GlInterface is wrong as I do not know where to get from so I store it internally on first load of the control in order to recall the methods on newly loaded data:

![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/2b477916-3a25-40c5-8086-c3a949c866db)

![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/442c6eff-43dc-4d53-8296-c4b1fc89e226)

To be honest I am not a 3D specialist so all the data filing and shader stuff has been created by our 3D guy ;)

**Reply by markusalbrecht-procam** [reply]
Initial Load:
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/fe207d65-6079-48da-a866-182d2e4ad876)

Newly loaded file:
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/2f5b6e8a-712c-415e-bf35-0f70f321c276)

**Comment by BykovskiyA** [comment]
Having tried to expand the capabilities of **OpenGlControlBase**, we encountered a problem that some OpenGL functions work correctly only when called inside **OpenGLInit**.
For our solution we need dynamically change the displayed 3D data.

Example of problem in pseudo code ():

```
----------------------------------------------------  
GlInterface gl;

Create3DDataBuffers()
{
    gl.BindBuffer(GL_ARRAY_BUFFER, _vertexBufferObject);
    fixed (void* pdata = _points)
        gl.BufferData(GL_ARRAY_BUFFER, new nint(_points.Length * vertexSize), new nint(pdata), GL_DYNAMIC_DRAW);

    gl.BindBuffer(GL_ELEMENT_ARRAY_BUFFER, _indexBufferObject);
    fixed (void* pdata = _indices)
        gl.BufferData(GL_ELEMENT_ARRAY_BUFFER, new nint(_indices.Length * sizeof(uint)), new nint(pdata), GL_DYNAMIC_DRAW);
}

OnOpenGlInit(GlInterface GL)

{
    gl = GL;
    ...

    // generate 3d data buffers
    // if called outside OnOpenGlInit - always return zero value
    _vertexBufferObject = gl.GenBuffer();
    _indexBufferObject = gl.GenBuffer();
    _vertexArrayObject = gl.GenVertexArray();

    ...

    Create3DDataBuffers();
}   
----------------------------------------------------  

```

However, we get unexpected behavior of some functions if the call occurs outside the function **OnOpenGlInit()**:
1) **gl.BindBuffer** - didn't recreate buffer data, the data remains as it was
2) **gl.GenBuffer()** - always return 0 (zero value)

Is this behavior of functions normal for Avalonia (some function calls work correctly only inside OnOpenGlInit())?

If this behavior is intended, can you tell me how we can change the data dynamically at any time after calling **OnOpenGlInit()**?
One option to work around the problems is to use **glBufferSubData()** (create several large buffers of 3D data and then change the contents of parts of that buffer), but this function has not been added.

**Reply by markusalbrecht-procam** [reply]
@kekekeks Hi, any ideas why this is not working or do you need more details ?

**Reply by BAndysc** [reply]
I believe this is _welcome to the OpenGL API problem_. Unlike Dx11 or Vulkan where you pass the context instance (vkInstance), OpenGL API always refers to the _current context_, which basically keeps all the state. https://www.khronos.org/opengl/wiki/OpenGL_Context OpenGlControlBase manages the context for you, the context is correctly set in `OnOpenGlInit`, `OnOpenGlRender` and  `OnOpenGlDeinit`, so you should use GL functions ONLY inside those methods. https://github.com/AvaloniaUI/Avalonia/blob/1c792b294f82b26396e4f91032a2b535ea9c0669/src/Avalonia.OpenGL/Controls/OpenGlControlBase.cs#L181 If you try to use any OpenGL function outside of those methods, they will not refer to _your control_ context and **not work** the way you want.

My tip is to schedule any work you need to do and execute the work inside the render method.

**Reply by markusalbrecht-procam** [reply]
@BAndysc YES that was it by recreating the Shader inside Rendering it updates with the newly loaded data !!!!
Thanks for that ;) - so this seems to be solved then.
